File: src/main/java/com/rokucraft/RokuBot/serializers/MessageEmbedSerializer.java
package com.rokucraft.RokuBot.serializers;

import com.rokucraft.RokuBot.Constants;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;
import org.spongepowered.configurate.serialize.TypeSerializer;

import java.lang.reflect.Type;

public class MessageEmbedSerializer implements TypeSerializer<MessageEmbed> {

    public static final MessageEmbedSerializer INSTANCE = new MessageEmbedSerializer();

    @Override
    public MessageEmbed deserialize(Type type, ConfigurationNode node) throws SerializationException {
        return new EmbedBuilder()
                .setTitle(node.node("title").getString(), node.node("url").getString())
                .setThumbnail(node.node("thumbnail-url").getString())
                .setDescription(node.node("description").getString())
                .setImage(node.node("image-url").getString())
                .setColor(node.node("color").getInt(Constants.BLUE))
                .setFooter(node.node("footer", "text").getString(), node.node("footer", "icon-url").getString())
                .build();
    }

    @Override
    public void serialize(Type type, @Nullable MessageEmbed obj, ConfigurationNode node) throws SerializationException {
        if (obj == null) {
            node.raw(null);
            return;
        }

        node.node("title").set(obj.getTitle());
        node.node("url").set(obj.getUrl());
        if (obj.getThumbnail() != null) {
            node.node("thumbnail-url").set(obj.getThumbnail().getUrl());
        }
        node.node("description").set(obj.getDescription());
        if (obj.getImage() != null) {
            node.node("image-url").set(obj.getImage().getUrl());
        }
        node.node("color").set(obj.getColorRaw());
        if (obj.getFooter() != null) {
            node.node("footer", "text").set(obj.getFooter().getText());
            node.node("footer", "icon-url").set(obj.getFooter().getIconUrl());
        }
    }
}

File: src/main/kotlin/config/serializers/MessageEmbedSerializer.kt
package com.rokucraft.rokubot.config.serializers

import com.rokucraft.rokubot.util.ColorConstants
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.MessageEmbed
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.kotlin.extensions.get
import org.spongepowered.configurate.serialize.SerializationException
import org.spongepowered.configurate.serialize.TypeSerializer
import java.lang.reflect.Type

object MessageEmbedSerializer : TypeSerializer<MessageEmbed?> {
    override fun deserialize(type: Type, node: ConfigurationNode): MessageEmbed {
        return EmbedBuilder()
            .setTitle(node.node("title").get(), node.node("url").get())
            .setThumbnail(node.node("thumbnail-url").get())
            .setDescription(node.node("description").get())
            .setImage(node.node("image-url").get())
            .setColor(node.node("color").getInt(ColorConstants.BLUE))
            .setFooter(node.node("footer", "text").get(), node.node("footer", "icon-url").get())
            .build()
    }

    @Throws(SerializationException::class)
    override fun serialize(type: Type, obj: MessageEmbed?, node: ConfigurationNode) {
        if (obj == null) {
            node.raw(null)
            return
        }
        node.node("title").set(obj.title)
        node.node("url").set(obj.url)
        if (obj.thumbnail != null) {
            node.node("thumbnail-url").set(obj.thumbnail!!.url)
        }
        node.node("description").set(obj.description)
        if (obj.image != null) {
            node.node("image-url").set(obj.image!!.url)
        }
        node.node("color").set(obj.colorRaw)
        if (obj.footer != null) {
            node.node("footer", "text").set(obj.footer!!.text)
            node.node("footer", "icon-url").set(obj.footer!!.iconUrl)
        }
    }
}
